File: standalone-go/valuetable.go
package main

// Loads the vendored BioScan-derived species value table and ExploData's genus display names.
// Same data as edexotracker's standalone/value_table.py, embedded directly into the binary via
// go:embed instead of read from disk at runtime -- see standalone/vendor/bio_value_data/README.md
// for full provenance/refresh instructions.

import (
	_ "embed"
	"encoding/json"
)

//go:embed vendor/value_table.json
var valueTableJSON []byte

//go:embed vendor/genus_names.json
var genusNamesJSON []byte

type speciesInfo struct {
	Name  string `json:"name"`
	Value int64  `json:"value"`
}

var valueTable map[string]map[string]speciesInfo
var genusNames map[string]string

func init() {
	if err := json.Unmarshal(valueTableJSON, &valueTable); err != nil {
		panic("failed to parse embedded value table: " + err.Error())
	}
	if err := json.Unmarshal(genusNamesJSON, &genusNames); err != nil {
		panic("failed to parse embedded genus names: " + err.Error())
	}
}

func lookupSpecies(genus, species string) (speciesInfo, bool) {
	if genus == "" || species == "" {
		return speciesInfo{}, false
	}
	sp, ok := valueTable[genus]
	if !ok {
		return speciesInfo{}, false
	}
	info, ok := sp[species]
	return info, ok
}

func genusDisplayName(genus string) string {
	if name, ok := genusNames[genus]; ok {
		return name
	}
	return genus
}

File: standalone/value_table.py
"""Loads standalone's own vendored species value table. See vendor/bio_value_data/README.md.
Deliberately identical interface to edexotracker/value_table.py, but reads standalone's own
copies -- no import dependency on the edexotracker package."""

from __future__ import annotations

import json
import os
from typing import NamedTuple, Optional

_VALUE_TABLE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'vendor', 'bio_value_data', 'value_table.json')
_GENUS_NAMES_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'vendor', 'bio_value_data', 'genus_names.json')


class SpeciesInfo(NamedTuple):
    name: str
    value: int


_TABLE: Optional[dict[str, dict[str, SpeciesInfo]]] = None


def load() -> dict[str, dict[str, SpeciesInfo]]:
    with open(_VALUE_TABLE_PATH) as f:
        raw = json.load(f)
    return {
        genus_key: {
            species_key: SpeciesInfo(name=info['name'], value=info['value'])
            for species_key, info in species_map.items()
        }
        for genus_key, species_map in raw.items()
    }


def lookup(genus: str, species: str) -> Optional[SpeciesInfo]:
    global _TABLE
    if _TABLE is None:
        _TABLE = load()
    if not genus or not species:
        return None
    return _TABLE.get(genus, {}).get(species)


_GENUS_NAMES: Optional[dict[str, str]] = None


def genus_display_name(genus: str) -> str:
    global _GENUS_NAMES
    if _GENUS_NAMES is None:
        with open(_GENUS_NAMES_PATH) as f:
            _GENUS_NAMES = json.load(f)
    return _GENUS_NAMES.get(genus, genus)
